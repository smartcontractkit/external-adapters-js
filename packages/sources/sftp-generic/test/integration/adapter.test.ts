import {
  TestAdapter,
  setEnvVariables,
} from '@chainlink/external-adapter-framework/util/testing-utils'
import { mockSftpClientInstance } from '../mocks/sftpClient'
import {
  ftse100CsvContent,
  getCurrentDateFormatted,
  getExpectedFileName,
  russell1000CsvContent,
  russell2000CsvContent,
  russell3000CsvContent,
} from './fixtures'

// Mock ssh2-sftp-client before importing the adapter
jest.mock('ssh2-sftp-client', () => {
  return require('../mocks/sftpClient').default
})

describe('SFTP Generic Adapter', () => {
  let spy: jest.SpyInstance
  let testAdapter: TestAdapter<any>
  let oldEnv: NodeJS.ProcessEnv

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))

    // Set required environment variables
    process.env.SFTP_HOST = 'localhost'
    process.env.SFTP_PORT = '2222'
    process.env.SFTP_USERNAME = 'testuser'
    process.env.SFTP_PASSWORD = 'testpass'
    process.env.BACKGROUND_EXECUTE_MS = '0'

    const mockDate = new Date('2024-08-23T10:00:00.000Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())

    const adapter = (await import('./../../src')).adapter
    adapter.rateLimiting = undefined
    testAdapter = (await TestAdapter.startWithMockedCache(adapter, {
      testAdapter: {} as TestAdapter<any>,
    })) as any
  })

  afterAll(async () => {
    setEnvVariables(oldEnv)
    await testAdapter.api.close()
    spy.mockRestore()
  })

  beforeEach(() => {
    // Reset mock state before each test
    mockSftpClientInstance.setFiles({})
    mockSftpClientInstance.setShouldFailConnection(false)
    mockSftpClientInstance.setShouldFailFileOperation(false)
    mockSftpClientInstance.setConnectionTimeout(false)
  })

  describe('sftp endpoint', () => {
    it('should fail with missing required parameters', async () => {
      const data = {
        endpoint: 'sftp',
      }
      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(400)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail with unsupported operation', async () => {
      const data = {
        endpoint: 'sftp',
        operation: 'upload',
        remotePath: '/data',
        instrument: 'FTSE100INDEX',
      }
      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(400)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail with unsupported instrument', async () => {
      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data',
        instrument: 'UNSUPPORTED_INDEX',
      }
      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(400)
      expect(response.json()).toMatchSnapshot()
    })

    it('should successfully download FTSE100INDEX file', async () => {
      const { day, month } = getCurrentDateFormatted()
      const fileName = getExpectedFileName('FTSE100INDEX', day, month)
      const filePath = `/data/${fileName}`

      // Set up mock files
      mockSftpClientInstance.setFiles({
        [filePath]: ftse100CsvContent,
      })

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data',
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })

    it('should successfully download Russell1000INDEX file', async () => {
      const { day, month } = getCurrentDateFormatted()
      const fileName = getExpectedFileName('Russell1000INDEX', day, month)
      const filePath = `/data/${fileName}`

      // Set up mock files
      mockSftpClientInstance.setFiles({
        [filePath]: russell1000CsvContent,
      })

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data',
        instrument: 'Russell1000INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })

    it('should successfully download Russell2000INDEX file', async () => {
      const { day, month } = getCurrentDateFormatted()
      const fileName = getExpectedFileName('Russell2000INDEX', day, month)
      const filePath = `/data/${fileName}`

      // Set up mock files
      mockSftpClientInstance.setFiles({
        [filePath]: russell2000CsvContent,
      })

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data',
        instrument: 'Russell2000INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })

    it('should successfully download Russell3000INDEX file', async () => {
      const { day, month } = getCurrentDateFormatted()
      const fileName = getExpectedFileName('Russell3000INDEX', day, month)
      const filePath = `/data/${fileName}`

      // Set up mock files
      mockSftpClientInstance.setFiles({
        [filePath]: russell3000CsvContent,
      })

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data',
        instrument: 'Russell3000INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail when file is not found', async () => {
      // Don't set up any files in the mock, and ensure mock will throw file not found error
      mockSftpClientInstance.setFiles({})

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data-notfound', // Use unique path to avoid cache hits
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(500)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail when SFTP connection fails', async () => {
      mockSftpClientInstance.setShouldFailConnection(true)

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data-connfail', // Use unique path to avoid cache hits
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(500)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail when SFTP connection times out', async () => {
      mockSftpClientInstance.setConnectionTimeout(true)

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data-timeout', // Use unique path to avoid cache hits
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(500)
      expect(response.json()).toMatchSnapshot()
    })

    it('should fail when file operation fails', async () => {
      mockSftpClientInstance.setShouldFailFileOperation(true)

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/data-opfail', // Use unique path to avoid cache hits
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(500)
      expect(response.json()).toMatchSnapshot()
    })

    it('should handle different remote paths correctly', async () => {
      const { day, month } = getCurrentDateFormatted()
      const fileName = getExpectedFileName('FTSE100INDEX', day, month)
      const filePath = `/custom/path/${fileName}`

      // Set up mock files
      mockSftpClientInstance.setFiles({
        [filePath]: ftse100CsvContent,
      })

      const data = {
        endpoint: 'sftp',
        operation: 'download',
        remotePath: '/custom/path',
        instrument: 'FTSE100INDEX',
      }

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    })
  })
})
