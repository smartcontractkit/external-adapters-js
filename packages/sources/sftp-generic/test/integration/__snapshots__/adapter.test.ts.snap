// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SFTP Generic Adapter sftp endpoint should fail when SFTP connection fails 1`] = `
{
  "errorMessage": "Failed to connect to SFTP server: Connection failed",
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data-connfail","instrument":"ftse100index"}",
    },
  },
  "statusCode": 500,
  "timestamps": {
    "providerDataReceivedUnixMs": 0,
    "providerDataRequestedUnixMs": 0,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail when SFTP connection times out 1`] = `
{
  "errorMessage": "Failed to connect to SFTP server: Connection timeout",
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data-timeout","instrument":"ftse100index"}",
    },
  },
  "statusCode": 500,
  "timestamps": {
    "providerDataReceivedUnixMs": 0,
    "providerDataRequestedUnixMs": 0,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail when file is not found 1`] = `
{
  "errorMessage": "Failed to download file: File not found: /data-notfound/vall2308.csv",
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data-notfound","instrument":"ftse100index"}",
    },
  },
  "statusCode": 500,
  "timestamps": {
    "providerDataReceivedUnixMs": 0,
    "providerDataRequestedUnixMs": 0,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail when file operation fails 1`] = `
{
  "errorMessage": "Failed to download file: File operation failed",
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data-opfail","instrument":"ftse100index"}",
    },
  },
  "statusCode": 500,
  "timestamps": {
    "providerDataReceivedUnixMs": 0,
    "providerDataRequestedUnixMs": 0,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail with missing required parameters 1`] = `
{
  "error": {
    "message": "[Param: operation] param is required but no value was provided",
    "name": "AdapterError",
  },
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail with unsupported instrument 1`] = `
{
  "errorMessage": "Unsupported instrument: UNSUPPORTED_INDEX",
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data","instrument":"unsupported_index"}",
    },
  },
  "statusCode": 400,
  "timestamps": {
    "providerDataReceivedUnixMs": 0,
    "providerDataRequestedUnixMs": 0,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should fail with unsupported operation 1`] = `
{
  "error": {
    "message": "[Param: operation] input is not one of valid options (download)",
    "name": "AdapterError",
  },
  "status": "errored",
  "statusCode": 400,
}
`;

exports[`SFTP Generic Adapter sftp endpoint should handle different remote paths correctly 1`] = `
{
  "data": {
    "result": "{"operation":"download","fileName":"vall2308.csv","path":"/custom/path","content":"Date,Open,High,Low,Close,Volume\\n2024-08-23,7500.00,7520.00,7485.00,7510.00,1000000\\n2024-08-22,7485.00,7505.00,7470.00,7500.00,950000\\n2024-08-21,7470.00,7490.00,7455.00,7485.00,900000","contentType":"text/csv","timestamp":1724407200000}",
  },
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/custom/path","instrument":"ftse100index"}",
    },
  },
  "result": "{"operation":"download","fileName":"vall2308.csv","path":"/custom/path","content":"Date,Open,High,Low,Close,Volume\\n2024-08-23,7500.00,7520.00,7485.00,7510.00,1000000\\n2024-08-22,7485.00,7505.00,7470.00,7500.00,950000\\n2024-08-21,7470.00,7490.00,7455.00,7485.00,900000","contentType":"text/csv","timestamp":1724407200000}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1724407200000,
    "providerDataRequestedUnixMs": 1724407200000,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should successfully download FTSE100INDEX file 1`] = `
{
  "data": {
    "result": "{"operation":"download","fileName":"vall2308.csv","path":"/data","content":"Date,Open,High,Low,Close,Volume\\n2024-08-23,7500.00,7520.00,7485.00,7510.00,1000000\\n2024-08-22,7485.00,7505.00,7470.00,7500.00,950000\\n2024-08-21,7470.00,7490.00,7455.00,7485.00,900000","contentType":"text/csv","timestamp":1724407200000}",
  },
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data","instrument":"ftse100index"}",
    },
  },
  "result": "{"operation":"download","fileName":"vall2308.csv","path":"/data","content":"Date,Open,High,Low,Close,Volume\\n2024-08-23,7500.00,7520.00,7485.00,7510.00,1000000\\n2024-08-22,7485.00,7505.00,7470.00,7500.00,950000\\n2024-08-21,7470.00,7490.00,7455.00,7485.00,900000","contentType":"text/csv","timestamp":1724407200000}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1724407200000,
    "providerDataRequestedUnixMs": 1724407200000,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should successfully download Russell1000INDEX file 1`] = `
{
  "data": {
    "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,2450.50,5000000,2500000000\\n2024-08-22,2440.25,4800000,2480000000\\n2024-08-21,2435.75,4600000,2460000000","contentType":"text/csv","timestamp":1724407200000}",
  },
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data","instrument":"russell1000index"}",
    },
  },
  "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,2450.50,5000000,2500000000\\n2024-08-22,2440.25,4800000,2480000000\\n2024-08-21,2435.75,4600000,2460000000","contentType":"text/csv","timestamp":1724407200000}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1724407200000,
    "providerDataRequestedUnixMs": 1724407200000,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should successfully download Russell2000INDEX file 1`] = `
{
  "data": {
    "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,1950.50,3000000,1950000000\\n2024-08-22,1940.25,2800000,1940000000\\n2024-08-21,1935.75,2600000,1930000000","contentType":"text/csv","timestamp":1724407200000}",
  },
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data","instrument":"russell2000index"}",
    },
  },
  "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,1950.50,3000000,1950000000\\n2024-08-22,1940.25,2800000,1940000000\\n2024-08-21,1935.75,2600000,1930000000","contentType":"text/csv","timestamp":1724407200000}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1724407200000,
    "providerDataRequestedUnixMs": 1724407200000,
  },
}
`;

exports[`SFTP Generic Adapter sftp endpoint should successfully download Russell3000INDEX file 1`] = `
{
  "data": {
    "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,2750.50,8000000,4450000000\\n2024-08-22,2740.25,7800000,4430000000\\n2024-08-21,2735.75,7600000,4410000000","contentType":"text/csv","timestamp":1724407200000}",
  },
  "meta": {
    "adapterName": "SFTP_GENERIC",
    "metrics": {
      "feedId": "{"operation":"download","remotePath":"/data","instrument":"russell3000index"}",
    },
  },
  "result": "{"operation":"download","fileName":"daily_values_russell_2308.csv","path":"/data","content":"Date,Price,Volume,MarketCap\\n2024-08-23,2750.50,8000000,4450000000\\n2024-08-22,2740.25,7800000,4430000000\\n2024-08-21,2735.75,7600000,4410000000","contentType":"text/csv","timestamp":1724407200000}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1724407200000,
    "providerDataRequestedUnixMs": 1724407200000,
  },
}
`;
